### PR TITLE
impl(oauth2): refactor ServiceAccountCredentials

### PR DESCRIPTION
**Pull request by scotthart**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9736)

### CODE REVIEW

File: google/cloud/storage/oauth2/service_account_credentials.cc [hunk 1]
```diff
@@ -136,14 +70,19 @@ std::string MakeJWTAssertion(std::string const& header,
...
-  return payload;
+  auto params = oauth2_internal::CreateServiceAccountRefreshPayload(
+      oauth2_internal::ServiceAccountCredentialsInfo{
+          info.client_email, info.private_key_id, info.private_key,
+          info.token_uri, info.scopes, info.subject},
+      now);
+  return absl::StrJoin(
+      params, "&", [](std::string* out, std::pair<std::string, std::string> const& h) {
+        std::cout << __PRETTY_FUNCTION__ << " " << h.first << ", " << h.second
+                  << std::endl;
```
Comment (devbww) [inline google/cloud/storage/oauth2/service_account_credentials.cc:83]: Debugging?

File: google/cloud/storage/oauth2/service_account_credentials.h [hunk 2]
```diff
@@ -254,6 +307,7 @@ class ServiceAccountCredentials : public Credentials {
   ClockType clock_;
 };
 
+
```
Comment (devbww) [inline google/cloud/storage/oauth2/service_account_credentials.h:310]: Remove extra vertical whitespace.

File: google/cloud/storage/oauth2/service_account_credentials.cc [hunk 3]
```diff
@@ -136,14 +70,19 @@ std::string MakeJWTAssertion(std::string const& header,
...
+        absl::StrAppend(out, h.first + "=" + h.second);
+      });
```
Comment (devbww) [inline google/cloud/storage/oauth2/service_account_credentials.cc:85]: Last argument looks like `absl::PairFormatter("=")`.

**Review by coryan — APPROVED** [review]
What @devbww said, but otherwise LGTM.

**Review by scotthart — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9736)* status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @devbww)

___
*[`google/cloud/storage/oauth2/service_account_credentials.h` line 310 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9736#-NAMQvxCSdy4m8zLVpSG-r1-310:-NAQCIde3GTGpyJtaxMl:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/81f508ebe779feec61110a3de883b6613cf09557/google/cloud/storage/oauth2/service_account_credentials.h#L310)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Remove extra vertical whitespace.
</blockquote></details>

Done.
___
*[`google/cloud/storage/oauth2/service_account_credentials.cc` line 83 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9736#-NAMQvxCSdy4m8zLVpSF-r1-83:-NAQCJUj53Ca7b_SOSxs:bss4a2b) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/81f508ebe779feec61110a3de883b6613cf09557/google/cloud/storage/oauth2/service_account_credentials.cc#L83)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Debugging?
</blockquote></details>

Removed.
___
*[`google/cloud/storage/oauth2/service_account_credentials.cc` line 85 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/9736#-NAMQvxCSdy4m8zLVpSF-r1-85:-NAQCKaWCabhmNWaF6tG:bhd3jeg) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/81f508ebe779feec61110a3de883b6613cf09557/google/cloud/storage/oauth2/service_account_credentials.cc#L85)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Last argument looks like `absl::PairFormatter("=")`.
</blockquote></details>

TIL: PairFormatter

**Pull request merged**